lib/Conversion/AffineToPipeline/AffineToPipeline.cpp

```diff
@@ -57,20 +57,73 @@ struct AffineToPipeline : public AffineToPipelineBase<AffineToPipeline> {
   LogicalResult populateOperatorTypes(SmallVectorImpl<AffineForOp> &loopNest);
   LogicalResult solveSchedulingProblem(SmallVectorImpl<AffineForOp> &loopNest);
   LogicalResult createPipelinePipeline(SmallVectorImpl<AffineForOp> &loopNest);
+  LogicalResult constantCSE();
+  LogicalResult unrollAndForwardStores();
 
   CyclicSchedulingAnalysis *schedulingAnalysis;
 };
 
+static LogicalResult
+forwardStoreToLoad(AffineReadOpInterface loadOp,
+                   SmallVectorImpl<Operation *> &opsToErase) {
+
+  Operation *prevOp = loadOp->getPrevNode();
+  while (prevOp) {
+    if (prevOp->getParentRegion() != loadOp->getParentRegion())
+      return failure();
+
+    auto storeOp = dyn_cast<AffineWriteOpInterface>(prevOp);
+    if (!storeOp) {
+      prevOp = prevOp->getPrevNode();
+      continue;
+    }
+
+    if (storeOp.getMemRef() != loadOp.getMemRef())
+      return failure();
+
+    MemRefAccess srcAccess(storeOp);
+    MemRefAccess dstAccess(loadOp);
+
+    if (srcAccess != dstAccess)
+      return failure();
+
+    // Perform the actual store to load forwarding.
+    Value storeVal = cast<AffineWriteOpInterface>(storeOp).getValueToStore();
+    // Check if 2 values have the same shape. This is needed for affine
+    // vector loads and stores.
+    if (storeVal.getType() != loadOp.getValue().getType())
+      return failure();
+    loadOp.getValue().replaceAllUsesWith(storeVal);
+    // Record this to erase later.
+    opsToErase.push_back(storeOp);
+    // Record this to erase later.
+    opsToErase.push_back(loadOp);
+    return success();
+  }
+  return failure();
+}
+
 } // namespace
 
 void AffineToPipeline::runOnOperation() {
-  // Get dependence analysis for the whole function.
-  auto dependenceAnalysis = getAnalysis<MemoryDependenceAnalysis>();
+  if (failed(unrollAndForwardStores()))
+    return signalPassFailure();
+
+  // auto dependenceAnalysis = getAnalysis<MemoryDependenceAnalysis>();
```

Comment:
Using this versus using hasNoIntervening effects is very different. I expect this to be very slow since this will try to check for every pair of load/store operations.

Removing this analysis for the whole function and doing it on a restricted subset (inside hasNoInterveningEffects) is a better idea.